code/datums/components/gravedigger.dm

```diff
@@ -0,0 +1,48 @@
+/**
+ * Gravedigger component. Allows for graves to be dug from certain tiles
+ */
+/datum/component/gravedigger
+	/// How long it takes to dig a grave
+	var/dig_time = 8 SECONDS
+	/// A list of turf types that can be used to dig a grave.
+	var/static/list/turfs_to_consider = typecacheof(list(
+		/turf/open/misc/asteroid,
+		/turf/open/misc/dirt,
+		/turf/open/misc/grass,
+		/turf/open/misc/basalt,
+		/turf/open/misc/ashplanet,
+		/turf/open/misc/snow,
+		/turf/open/misc/sandy_dirt,
+	))
+
+/datum/component/gravedigger/Initialize(
+	dig_time = 8 SECONDS,
+)
+
+	if(!isobj(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	src.dig_time = dig_time
+
+/datum/component/gravedigger/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ITEM_INTERACTING_WITH_ATOM_SECONDARY, PROC_REF(dig_checks))
+
+/datum/component/gravedigger/proc/dig_checks(datum/source, mob/living/user, atom/interacting_with, list/modifiers)
+	SIGNAL_HANDLER
+
+	if(!is_type_in_typecache(interacting_with, turfs_to_consider))
+		return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
```

Comment:
If it's not in the typecache, you probably want to return `NONE`, ie, do normal handling, rather than cancelling